### PR TITLE
Matrix multiplication tutorial block pointer variant

### PR DESCRIPTION
**Pull request by rolfmorel**
This is on top of the `dev-cpu-v2` branch, so that we can compare the effect of the changed IR for that approach as well.

### CODE REVIEW

File: python/tutorials/03-matrix-multiplication-cpu.py [hunk 1]
```diff
@@ -244,6 +245,65 @@ def matmul_kernel(
...
+        block_shape=(BLOCK_SIZE_M, BLOCK_SIZE_K),
+        order=(1, 0),
+    )
+    b_tile_ptr = tl.make_block_ptr(
+        base=b_ptr,
+        shape=(K, N),
+        strides=(stride_bk, stride_bn),
+        offsets=(0, block_offset_n),
+        block_shape=(BLOCK_SIZE_K, BLOCK_SIZE_N),
+        order=(1, 0),
```
Comment (adam-smnk) [inline python/tutorials/03-matrix-multiplication-cpu.py:290]: Should all these `order` fields be this way?

Comment (rolfmorel) [reply python/tutorials/03-matrix-multiplication-cpu.py:290]: The code for this largely comes from one of the test case: https://github.com/plaidml/triton-cpu/blob/f30e7aafc014e6ba482782128afc9c3ec3ab040e/python/test/unit/hopper/test_gemm.py#L144

There the order is determined through parameters: https://github.com/plaidml/triton-cpu/blob/f30e7aafc014e6ba482782128afc9c3ec3ab040e/python/test/unit/hopper/test_gemm.py#L177

These parameters are `[1, 0]` in the non-transposed case, for all three matrices: https://github.com/plaidml/triton-cpu/blob/f30e7aafc014e6ba482782128afc9c3ec3ab040e/python/test/unit/hopper/test_gemm.py#L357

A further sanity check is that when run the output is identical to Torch's output for the random matmul.

File: python/tutorials/03-matrix-multiplication-cpu.py [hunk 2]
```diff
@@ -244,6 +245,65 @@ def matmul_kernel(
...
+
+    for k in range(0, K, BLOCK_SIZE_K):
+        a = tl.load(a_tile_ptr, boundary_check=(0, 1))
+        b = tl.load(b_tile_ptr, boundary_check=(0, 1))
+        accumulator += tl.dot(a, b)
+        a_tile_ptr = tl.advance(a_tile_ptr, [0, BLOCK_SIZE_K])
+        b_tile_ptr = tl.advance(b_tile_ptr, [BLOCK_SIZE_K, 0])
+
+    c = accumulator
+    c_block_ptr = tl.make_block_ptr(base=c_ptr, shape=(M, N), strides=(stride_cm, stride_cn),
```
Comment (rengolin) [inline python/tutorials/03-matrix-multiplication-cpu.py:302]: nit: indent like the other two to make comparisons easier

**Review by rengolin — COMMENTED** [review]
This looks reasonable to me. And it doesn't assume block pointers input, which is good.

With the two different styles present, it will be easy to check output and performance side-to-side.

File: python/tutorials/03-matrix-multiplication-cpu.py (hunk 1 above)
Comment (rengolin) [reply python/tutorials/03-matrix-multiplication-cpu.py:290]: Does that mean both A and B tiles get a transpose?

Comment (adam-smnk) [reply python/tutorials/03-matrix-multiplication-cpu.py:290]: There's a test that checks outputs against native torch execution. Could you run it with block ptrs for validation?

Comment (adam-smnk) [reply python/tutorials/03-matrix-multiplication-cpu.py:290]: > A further sanity check is that when run the output is identical to Torch's output for the random matmul.

Ahh sorry skipped over this. If it runs fine then cool.

Comment (adam-smnk) [reply python/tutorials/03-matrix-multiplication-cpu.py:290]: > Does that mean both A and B tiles get a transpose?

Looking at the `test_gemm.py` seems that `[1, 0]` marks no transposition, maybe it's just coming from column major convention.
@rolfmorel maybe you could look at maps of lowered vector.contract?

Comment (rolfmorel) [reply python/tutorials/03-matrix-multiplication-cpu.py:290]: Here's the IR with the relevant maps. `%arg0` is A and `%arg1` is B.
```
#map = affine_map<(d0, d1, d2) -> (d0, d2)>
#map1 = affine_map<(d0, d1, d2) -> (d2, d1)>
#map2 = affine_map<(d0, d1, d2) -> (d0, d1)>
...
    %19 = tt.make_tensor_ptr %arg0, [%16, %17], [%18, %c1_i64], [%14, %c0_i32] {order = array<i32: 1, 0>} : <tensor<16x16xf32>> loc(#loc18)
    %20 = arith.extsi %arg4 : i32 to i64 loc(#loc19)
    %21 = arith.extsi %arg7 : i32 to i64 loc(#loc19)
    %22 = tt.make_tensor_ptr %arg1, [%17, %20], [%21, %c1_i64], [%c0_i32, %15] {order = array<i32: 1, 0>} : <tensor<16x16xf32>> loc(#loc19)
    %23:3 = scf.for %arg9 = %c0_i32 to %arg5 step %c16_i32 iter_args(%arg10 = %cst_0, %arg11 = %19, %arg12 = %22) -> (vector<16x16xf32>, !tt.ptr<tensor<16x16xf32>>, !tt.ptr<tensor<16x16xf32>>)  : i32 {
      %28 = triton_cpu.extract_memref %arg11 : <tensor<16x16xf32>> -> memref<?x?xf32, strided<[?, 1]>> loc(#loc21)
      %29:2 = triton_cpu.extract_indices %arg11 : <tensor<16x16xf32>> -> index, index loc(#loc21)
      %30 = vector.transfer_read %28[%29#0, %29#1], %cst : memref<?x?xf32, strided<[?, 1]>>, vector<16x16xf32> loc(#loc21)
      %31 = triton_cpu.extract_memref %arg12 : <tensor<16x16xf32>> -> memref<?x?xf32, strided<[?, 1]>> loc(#loc22)
      %32:2 = triton_cpu.extract_indices %arg12 : <tensor<16x16xf32>> -> index, index loc(#loc22)
      %33 = vector.transfer_read %31[%32#0, %32#1], %cst : memref<?x?xf32, strided<[?, 1]>>, vector<16x16xf32> loc(#loc22)
      %34 = vector.contract {indexing_maps = [#map, #map1, #map2], iterator_types = ["parallel", "parallel", "reduction"], kind = #vector.kind<add>} %30, %33, %arg10 : vector<16x16xf32>, vector<16x16xf32> into vector<16x16xf32> loc(#loc23)
      %35 = tt.advance %arg11, [%c0_i32, %c16_i32] : <tensor<16x16xf32>> loc(#loc24)
      %36 = tt.advance %arg12, [%c16_i32, %c0_i32] : <tensor<16x16xf32>> loc(#loc25)
      scf.yield %34, %35, %36 : vector<16x16xf32>, !tt.ptr<tensor<16x16xf32>>, !tt.ptr<tensor<16x16xf32>> loc(#loc26)
    } loc(#loc20)
 ```

Comment (adam-smnk) [reply python/tutorials/03-matrix-multiplication-cpu.py:290]: Looks good in terms of ops.
Do you know why it produces dynamically shaped memref? I don't think our XSMM code can handle that.

File: python/tutorials/03-matrix-multiplication-cpu.py [hunk 3]
```diff
@@ -218,31 +241,111 @@ def matmul_kernel(
...
+        c_ptrs = c_ptr + stride_cm * offs_cm[:, None] + stride_cn * offs_cn[None, :]
+
+        # TODO: Currently masked load is not supported yet.
+        # c_mask = (offs_cm[:, None] < M) & (offs_cn[None, :] < N)
+        # tl.store(c_ptrs, c, mask=c_mask)
+        tl.store(c_ptrs, c)
+
+
+@triton.jit
+def matmul_kernel_block_ptr(
```
Comment (rolfmorel) [inline python/tutorials/03-matrix-multiplication-cpu.py:292]: Still need to test modified `matmul_kernel` above before removing this.

(Testing is being made difficult due me installing triton-shared and now I am unable to uninstall it/install over it...)

File: python/tutorials/03-matrix-multiplication-cpu.py (hunk 2 above)
Comment (rolfmorel) [reply python/tutorials/03-matrix-multiplication-cpu.py]: Done.

**Comment by rolfmorel** [comment]
I changed the PR so that whether to use block-pointers or causes different versions of `matmul_kernel` to be generated based on its new "meta-parameter" `USE_BLOCK_POINTERS`.

@rengolin & @adam-smnk , if you could have a final look, that would be great.

File: python/tutorials/03-matrix-multiplication-cpu.py (hunk 3 above)
Comment (rolfmorel) [reply python/tutorials/03-matrix-multiplication-cpu.py]: Have confirmed the new version works fine, i.e. still generates the same IR.

File: python/tutorials/03-matrix-multiplication-cpu.py [hunk 4]
```diff
@@ -218,37 +241,56 @@ def matmul_kernel(
...
+        # TODO: masking
+        c_block_ptr = tl.make_block_ptr(
+            base=c_ptr,
+            shape=(M, N),
+            strides=(stride_cm, stride_cn),
+            offsets=(block_offset_m, block_offset_n),
+            block_shape=(BLOCK_SIZE_M, BLOCK_SIZE_N),
+            order=(1, 0)
+        )
+        tl.store(c_block_ptr, accumulator, boundary_check=(0, 1))
```
Comment (adam-smnk) [inline python/tutorials/03-matrix-multiplication-cpu.py:277]: nit: shouldn't it use `c` instead of `accumulator`?

**Review by adam-smnk — APPROVED** [review]

File: python/tutorials/03-matrix-multiplication-cpu.py (hunk 4 above)
Comment (rolfmorel) [reply python/tutorials/03-matrix-multiplication-cpu.py]: Indeed, to keep it in line with the existing example. Not sure what the benefit of it is though.

**Pull request merged**